src/v/raft/consensus.cc

```diff
@@ -2478,6 +2478,27 @@ model::offset consensus::get_latest_configuration_offset() const {
     return _configuration_manager.get_latest_offset();
 }
 
+result<bool> consensus::is_under_replicated(model::node_id node_id) const {
+    if (!is_leader()) {
+        return errc::not_leader;
+    }
+    auto it = std::find_if(
+      _fstats.begin(), _fstats.end(), [node_id](const auto& s) {
+          return s.first.id() == node_id;
+      });
+    if (it == _fstats.end()) {
+        return errc::node_does_not_exists;
+    }
+
+    auto dirty_offset = _log.offsets().dirty_offset;
```

Comment:
I'm not sure the dirty_index check makes sense here.

When we make a comparison like this in update_follower_index->needs_recovery, the dirty index we're comparing with is the value as it was when we dispatched AppendEntries, not the instantaneous value.

I think doing this check here is going to fail whenever writes are in flight that haven't been acked by the follower yet.

Maybe it makes sense to do this comparison with `committed_offset` instead of `dirty_offset`?

Reply:
This is basically copy-pasted from `get_follower_metrics()`, just below, I should probably factor it out, or just call `get_follower_metrics().under_replicated`

Reply:
Oh right, the offset check only kicks in if the node is already behind on heartbeats or marked recovering.